core/base/workspace_aliases.hpp

```diff
@@ -0,0 +1,255 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_BASE_WORKSPACE_ALIASES_HPP_
+#define GKO_CORE_BASE_WORKSPACE_ALIASES_HPP_
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+// This code is a modified version of the code from CCCL
+// (https://github.com/NVIDIA/cccl) (cub/detail/temporary_storage.cuh and
+// cub/temporary_storage.cuh), made available through the Apache-2.0 and BSD-3
+// licenses. See ABOUT-LICENSING.md for more details.
+
+
+namespace gko {
+namespace detail {
+
+
+template <int num_allocs, int align_bytes = 8>
+GKO_ATTRIBUTES GKO_INLINE GKO_DEVICE_ERROR_TYPE create_workspace_aliases(
+    void* workspace_ptr, size_t& num_bytes, void* (&allocations)[num_allocs],
+    size_t (&allocation_sizes)[num_allocs])
+{
+    constexpr int align_mask = ~(align_bytes - 1);
+
+    // Compute exclusive prefix sum over allocation requests
+    size_t allocation_offsets[num_allocs];
+    size_t bytes_needed = 0;
+    for (int i = 0; i < num_allocs; ++i) {
+        size_t allocation_bytes =
+            (allocation_sizes[i] + align_bytes - 1) & align_mask;
+        allocation_offsets[i] = bytes_needed;
+        bytes_needed += allocation_bytes;
+    }
+    bytes_needed += align_bytes - 1;
+
+    // Check if the caller is simply requesting the size of the storage
+    // allocation
+    if (!workspace_ptr) {
+        num_bytes = bytes_needed;
+        return GKO_DEVICE_NO_ERROR;
+    }
+
+    // Check if enough storage provided
+    if (num_bytes < bytes_needed) {
+        return GKO_DEVICE_ERROR_INVALID;
+    }
+
+    // Alias
+    workspace_ptr =
+        (void*)((size_t(workspace_ptr) + align_bytes - 1) & align_mask);
+    for (int i = 0; i < num_allocs; ++i) {
+        allocations[i] =
+            static_cast<char*>(workspace_ptr) + allocation_offsets[i];
+    }
+
+    return GKO_DEVICE_NO_ERROR;
+}
+
+
+class slot;
+
+template <typename T>
+class alias;
+
+template <int num_slots, int align_bytes>
+class layout;
+
+/**
+ * Represents a slot of memory which can be aliased. See @alias and @layout on
+ * how this can be used.
+ */
+class slot {
```

Comment:
still no idea why there is both `slot` and `alias`. `alias` alone should be enough. But I won't hold anything up over the implementations in this file.